src/rulesets/Base/array.jl

```diff
@@ -278,3 +278,128 @@ function rrule(::typeof(fill), value::Any, dims::Int...)
     end
     return fill(value, dims), fill_pullback
 end
+
+#####
+##### `findmax`, `maximum`, etc.
+#####
+
+for findm in (:findmin, :findmax)
+    findm_pullback = Symbol(findm, :_pullback)
+
+    @eval function frule((_, xdot), ::typeof($findm), x; dims=:)
+        y, ind = $findm(x; dims=dims)
+        return (y, ind), Tangent{typeof((y, ind))}(xdot[ind], NoTangent())
+    end
+
+    @eval function rrule(::typeof($findm), x::AbstractArray{<:Number}; dims=:)
+        y, ind = $findm(x; dims=dims)
+        project = ProjectTo(x)
+        function $findm_pullback((dy, _))
+            x_thunk = @thunk begin
+                dx = fill!(similar(x, eltype(dy)), false)
+                view(dx, ind) .= dy  # possibly 0-dim view, allows dy::Number and dy::Array, and dx::CuArray
+                project(dx)
+            end
+            x_ithunk = InplaceableThunk(x_thunk) do dx
+                view(dx, ind) .= view(dx, ind) .+ dy  # this could be .+=, but not on Julia 1.0
+                dx
+            end
+            return (NoTangent(), x_ithunk)
+        end
```

Comment:
could we close over size of x only here?

alternatively, I wonder whether we could reuse the rrule for getindex?
i.e. something like
```suggestion
        function $findm_pullback((dy, _))
            _, getindex_back = rrule(getindex, x, ind)
            return getindex_back(dy)[1:2]
        end
```

Reply:
> could we close over size of x only here?

There is `similar(typeof(x), size(x))` but no `similar(typeof(x), T, size(x))`. Of course very often ProjectTo is going to ensure that the eltype should really be the same as `x`'s, but not quite always. It's a bit awkward.

`getindex` has `@thunk(getindex_add!(zero(x)))` which seems worse -- it is not always going to be mutable, and it won't handle structure well, e.g. `zero(Diagonal([1,2,3]))[1,2] = 9`.

Reply:
I am not sure I understand the `getindex` comment. Isn't `similar` also going to make a `Diagonal`, just like `zero` does?

Reply:
Oh yea, both mess that one up, you need `similar(Diagonal(rand(3)), Int, (3,3))` to get something sure to be writeable. It's `zero(SA[1,2,3])` which is worse than `similar`.

Reply:
>Or maybe InplaceableStuff will make that obsolete anyway.

Is that your PR which makes InplaceableThunk take the third argument? I like that idea

Reply:
Oh maybe that too. But I meant that `Zygote.OneElement` is part of trying to speed up scalar indexing in a loop. The next step is https://github.com/FluxML/Zygote.jl/pull/981 . But this is a bit of a Zygote-style hack, and the eventual version involve ChainRules's in-place stuff. Then a tight loop might generate a million thunks, instead of a million `OneElement` "arrays".